grudge/trace_pair.py

```diff
@@ -272,61 +277,130 @@ def connected_ranks(dcoll: DiscretizationCollection):
     return get_connected_partitions(dcoll._volume_discr.mesh)
 
 
+def _flatten_to_numpy_from_container(
+        actx: ArrayContext, ary: ArrayOrContainerT) -> ArrayT:
+    """Flattens an arbitrary array container *ary* into a one-dimensional
+    numpy array suitable for communication via mpi4py.
+    """
+    if is_array_container(ary):
+        flat_data = []
+        # NOTE: The ordering of data in the flattened array will be determined
+        # by the ordering of key/value pairs returned by *serialize_container*.
+        # This is important to remember because the function
+        # *_unflatten_from_numpy_to_container* relies on this same ordering
+        # to unflatten the concatenated array.
+        for _, value in serialize_container(to_numpy(ary, actx)):
```

Comment:
In order to be valid, this needs an addition to the array container serialization spec (make a PR), namely that the iterables come out in the same order every time, for types of the same cardinality. I'm not opposed to that assumption, but it's needed in order for this to be correct, and it's not currently there.

Reply:
@inducer I quickly scanned through `arraycontext` to get a better understanding of how `serialize_container` is implemented to try to see if it is straightforward to extend. It's not... (to me)

Instead, I opted to just simply use an OrderedDict as a data template and perform reconstructions based on that: https://github.com/inducer/grudge/pull/154/commits/9efbe2b3b3f957fe8ae303a94186ffd6f1fa0034

Is it a good solution? Well, it seems to work just fine. I don't think it's any worse than what's already floating around. But perhaps you have a different opinion on this?

Reply:
I'm not sure I understand your comment here. What I was asking for was a PR into arraycontext to document that `serialize_container` produces the keys in a reliable order. If that's not the case, you can't assume that you got the keys for the template in the same order as the keys for the original data. (I.e. no code change, just codify a property that all current implementations already satisfy.)

Reply:
Sorry. I completely misunderstood what you were originally asking.